mocks/smart_mock_io.cpp

```diff
@@ -171,18 +172,51 @@ void
 bzn::asio::smart_mock_io::do_incoming_connection(size_t id)
 {
     this->tcp_accept_handlers.at(id)(boost::system::error_code{});
+    this->yield_until_clear();
     this->ws_accept_handlers.at(id)(boost::system::error_code{});
+    this->yield_until_clear();
 }
 
 void
 bzn::asio::smart_mock_io::shutdown()
 {
+    if (this->pending_real_callbacks > 0)
+    {
+        LOG(error) << "Warning: shutting down mock io context while it still has pending real callbacks to execute";
+    }
+
     // These callbacks are likely to transitively hold a shared pointers to us,
     // so cleaning them up is necessary for us to be cleaned up
     this->timer_callbacks.clear();
     this->ws_read_closures.clear();
     this->ws_write_closures.clear();
     this->ws_accept_handlers.clear();
     this->tcp_accept_handlers.clear();
+}
 
+void
```

Comment:
cleaning up outstanding callbacks. I can understand how this might trigger valgrind